extendedcopy.go

```diff
@@ -241,62 +252,90 @@ func (opts *ExtendedCopyGraphOptions) FilterArtifactType(regex *regexp.Regexp) {
 	if regex == nil {
 		return
 	}
+	filter := func(filtered []ocispec.Descriptor, predecessors []ocispec.Descriptor) []ocispec.Descriptor {
+		for _, p := range predecessors {
+			if regex.MatchString(p.ArtifactType) {
+				filtered = append(filtered, p)
+			}
+		}
+		return filtered
+	}
+
 	fp := opts.FindPredecessors
 	opts.FindPredecessors = func(ctx context.Context, src content.ReadOnlyGraphStorage, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
 		var predecessors []ocispec.Descriptor
 		var err error
+		referrerFinder, supportReferrers := src.(registry.ReferrerFinder)
 		if fp == nil {
-			// if src is a ReferrerFinder, use Referrers() to filter the predecessors.
-			if rf, ok := src.(registry.ReferrerFinder); ok {
-				return findReferrersAndFilter(rf, ctx, desc, regex)
+			if supportReferrers {
+				// if src is a ReferrerFinder, use Referrers() for possible memory saving
+				if err := referrerFinder.Referrers(ctx, desc, "", func(referrers []ocispec.Descriptor) error {
+					// for each page of the results, filter the referrers
+					predecessors = filter(predecessors, referrers)
+					return nil
+				}); err != nil {
+					return nil, err
+				}
+				return predecessors, nil
 			}
+
 			predecessors, err = src.Predecessors(ctx, desc)
 		} else {
 			predecessors, err = fp(ctx, src, desc)
 		}
 		if err != nil {
 			return nil, err
 		}
-		var filtered []ocispec.Descriptor
-		// for each predecessor, decode the manifest and check its artifact type.
-		for _, p := range predecessors {
-			if p.MediaType == artifactspec.MediaTypeArtifactManifest {
-				if err = func() error {
-					rc, err := src.Fetch(ctx, p)
+
+		if !supportReferrers {
+			// For src that does not support Referrers, Predecessors is used.
+			// However, the descriptors returned by Predecessors may not include
+			// the artifact type of the corresponding manifests.
+			for i, p := range predecessors {
+				if p.MediaType == ocispec.MediaTypeArtifactManifest && p.ArtifactType == "" {
```

Comment:
We might need to review the spec. For OCI image manifest, the config media type is used as artifact type.